### PR TITLE
nix doctor: add command

### PR DESCRIPTION
**Pull request by LnL7**
This command performs a bunch of sanity checks on the nix installation, most of these are not necceceraly a problem but can be very useful when debugging an issue. Most notably a way to detect worker/serve protocol mismatches in various conditions. This also works in combination with `--store ssh://host` to help debug eg. nix-copy-closure issues like https://github.com/NixOS/nix/pull/2383.

TODO: also include the details `nix-info` generates?

The code probably also needs a bit of cleanup.

---

```
$ nix doctor
Store uri: daemon

Warning: multiple versions of nix found in PATH.

  /nix/store/1wgsprdh6lafnqf8iddnm01k9kxvx5xv-nix-2.0.4/bin
  /src/nix/inst/bin

Warning: found profiles outside of /nix/var/nix/profiles.
The generation this profile points to might not have a gcroot and could be
garbage collected, resulting in broken symlinks.

  /tmp/foo/bin

Warning: protocol version of this client does not match the store.
While this is not necessarily a problem it's recommended to keep the client in
sync with the daemon.

Client protocol: 1.21
Store protocol: 1.20
```

### CODE REVIEW

**Comment by LnL7** [comment]
I didn't spend a lot of time on the error messages, they can definitively be improved. But I'm not sure if moving the all the `<profile>` with all of it's `<profile>-<gen>-link` symlinks and creating a pointer to those (like ~/.nix-profile) is always correct.

File: src/nix/doctor.cc [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                auto profileDir = canonPath(dirOf(dir), true);
+                if (hasSuffix(profileDir, "user-environment") &&
+                    store->isValidPath(profileDir)) {
+                    PathSet referrers;
+                    store->computeFSClosure({profileDir}, referrers, true,
+                            settings.gcKeepOutputs, settings.gcKeepDerivations);
+                    bool found = false;
+                    for (auto & i : roots)
+                        if (referrers.find(i.second) != referrers.end())
+                            found = true;
```
Comment (LnL7) [inline src/nix/doctor.cc]: Is there a better way to do this? It's a slow and can trigger gc unrelated messages.

**Comment by copumpkin** [comment]
This is awesome. As the go-to person for "dude my Nix is broken halp" at my company, this will be a godsend for me 😄

**Comment by copumpkin** [comment]
Not sure how best to implement it, but some way to determine if `nix-daemon` was started before your changed nix.conf would be pretty neat. I've seen a lot of people get confused because they change a setting in nix.conf and don't realize they need to restart the daemon.

**Comment by timokau** [comment]
Would it make sense to merge this with `nix-info`? That way you could just tell somebody to execute `nix doctor` and post the results (if they can't fix it themselves based on the report).

**Comment by tomberek** [comment]
Merge?

**Pull request merged**